### PR TITLE
feat: result instead of exit(1) on trap in recursion

### PR DESCRIPTION
**Pull request by mattstam**
bubble up `SP1RecursionProverError` instead of exit(1)

### CODE REVIEW

File: prover/src/lib.rs [hunk 1]
```diff
@@ -482,7 +477,7 @@ impl<C: SP1ProverComponents> SP1Prover<C> {
             let batched_compress_inputs =
                 compress_inputs.chunks(shard_batch_size).collect::<Vec<_>>();
             reduce_proofs = batched_compress_inputs
-                .into_iter()
+                .into_par_iter()
```
Comment (mattstam) [inline prover/src/lib.rs:480]: this seems kind of free? Or is there a reason we chose to not do this concurrently? just lmk and I'll remove

Comment (jtguibas) [reply prover/src/lib.rs:480]: Ah no, good catch. That looks good :)

File: prover/src/lib.rs [hunk 2]
```diff
@@ -538,11 +533,15 @@ impl<C: SP1ProverComponents> SP1Prover<C> {
         witness_stream.extend(input.write());
 
         runtime.witness_stream = witness_stream.into();
-        runtime.run();
```
Comment (jtguibas) [inline prover/src/lib.rs:541]: more canonical way to do this would be to do runtime.run().map_err(..)?;

File: prover/src/lib.rs [hunk 3]
```diff
@@ -579,7 +580,11 @@ impl<C: SP1ProverComponents> SP1Prover<C> {
...
+
```
Comment (jtguibas) [inline prover/src/lib.rs:583]: same here

File: prover/src/lib.rs [hunk 4]
```diff
@@ -579,7 +580,11 @@ impl<C: SP1ProverComponents> SP1Prover<C> {
...
+        if let Err(e) = runtime.run() {
+            return Err(SP1RecursionProverError::RuntimeError(e.to_string()));
```
Comment (jtguibas) [inline prover/src/lib.rs:585]: same

File: recursion/circuit/src/constraints.rs [hunk 5]
```diff
@@ -294,7 +294,7 @@ mod tests {
         let program = basic_program::<F>();
         let config = SC::new();
         let mut runtime = Runtime::<F, EF, DiffusionMatrixBabyBear>::new_no_perm(&program);
-        runtime.run();
```
Comment (jtguibas) [inline recursion/circuit/src/constraints.rs:297]: unwrap

File: recursion/compiler/examples/fibonacci.rs [hunk 6]
```diff
@@ -52,7 +52,7 @@ fn main() {
 
     let config = SC::new();
     let mut runtime = Runtime::<F, EF, _>::new(&program, config.perm.clone());
-    runtime.run();
```
Comment (jtguibas) [inline recursion/compiler/examples/fibonacci.rs:55]: unwrap

File: recursion/compiler/src/ir/utils.rs [hunk 7]
```diff
@@ -326,7 +326,7 @@ mod tests {
         let program = builder.compile_program();
 
         let mut runtime = Runtime::<F, EF, _>::new(&program, config.perm.clone());
```
Comment (jtguibas) [inline recursion/compiler/src/ir/utils.rs:328]: unrwap

File: prover/src/lib.rs (hunk 2 above)
Comment (mattstam) [reply prover/src/lib.rs:541]: updated

File: prover/src/lib.rs (hunk 3 above)
Comment (mattstam) [reply prover/src/lib.rs:583]: updated

File: prover/src/lib.rs (hunk 4 above)
Comment (mattstam) [reply prover/src/lib.rs]: updated

File: recursion/circuit/src/constraints.rs (hunk 5 above)
Comment (mattstam) [reply recursion/circuit/src/constraints.rs:297]: updated

File: recursion/compiler/examples/fibonacci.rs (hunk 6 above)
Comment (mattstam) [reply recursion/compiler/examples/fibonacci.rs:55]: updated

File: recursion/compiler/src/ir/utils.rs (hunk 7 above)
Comment (mattstam) [reply recursion/compiler/src/ir/utils.rs:328]: updated

**Review by jtguibas — APPROVED** [review]

**Pull request merged**